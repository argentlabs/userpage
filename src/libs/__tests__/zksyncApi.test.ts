import fetchMock from "jest-fetch-mock"

import {
  Config,
  ResultConfig,
  TokensZkSync,
  fetchConfig,
  fetchTokenList,
} from "../zksyncApi"
import {
  configApiResponseMock,
  tokensApiResponseMock,
} from "./__mocks__/zksyncApi.mock"

describe("[unit] ZKSYNC Api", () => {
  beforeAll(() => {
    fetchMock.enableMocks()
  })
  beforeEach(() => {
    fetchMock.resetMocks()
  })
  afterAll(() => {
    fetchMock.disableMocks()
  })
  describe("[unit] fetchConfig", () => {
    it("transforms 200", async () => {
      const mockResponse: Config = configApiResponseMock
      const expectedResult: ResultConfig = configApiResponseMock.result
      fetchMock.mockResponseOnce(JSON.stringify(mockResponse))

      const response = await fetchConfig()

      expect(response).toStrictEqual(expectedResult)
    })
    it("throws on 404", async () => {
      fetchMock.mockOnce(async () => ({ status: 404, body: "" }))

      await expect(fetchConfig).rejects.toThrowError()
    })
    it("throws on 500", async () => {
      fetchMock.mockOnce(async () => ({ status: 500, body: "" }))

      await expect(fetchConfig).rejects.toThrowError()
    })
  })
  describe("[unit] fetchTokenList", () => {
    it("transforms 200", async () => {
      const mockResponse: TokensZkSync = tokensApiResponseMock
      const expectedResult = new Set(tokensApiResponseMock.result.list)
      fetchMock.mockResponseOnce(JSON.stringify(mockResponse))

      const response = await fetchTokenList()

      // ignore token order
      expect(new Set(response)).toEqual(expectedResult)
    })
    it("throws on 404", async () => {
      fetchMock.mockOnce(async () => ({ status: 404, body: "" }))

      await expect(fetchConfig).rejects.toThrowError()
    })
    it("throws on 500", async () => {
      fetchMock.mockOnce(async () => ({ status: 500, body: "" }))

      await expect(fetchConfig).rejects.toThrowError()
    })
  })
})

describe("[int] ZKSYNC Api", () => {
  describe("[int] fetchConfig", () => {
    it("fetches", async () => {
      const fetchReal = fetch

      fetchMock.enableMocks()

      const expectedResult: ResultConfig = configApiResponseMock.result

      let lastResponseClone: Response | undefined
      const fetchJestFn = jest.fn<
        Promise<Response>,
        [input: RequestInfo, init?: RequestInit | undefined]
      >((...args) =>
        fetchReal(...args).then((x) => {
          lastResponseClone = x.clone()
          return x
        }),
      )
      fetchMock.mockImplementation(fetchJestFn as any)

      const response = await fetchConfig()

      expect(response).toStrictEqual(expectedResult)
      // fetch gets called once
      expect(fetchJestFn.mock.calls.length).toBe(1)
      // calls the right url
      expect(fetchJestFn.mock.calls[0]).toEqual([
        "https://ropsten-beta-api.zksync.io/api/v0.2/config",
      ])
      // gets correct api response
      const fetchResponse: Response = await fetchJestFn.mock.results[0].value
      expect(fetchResponse.status).toBe(200)
      expect((await lastResponseClone?.json()).result).toMatchSnapshot(
        "fetchConfig",
      )

      fetchMock.disableMocks()
    })
  })
  describe("[int] fetchTokenList", () => {
    it("fetches", async () => {
      const fetchReal = fetch

      fetchMock.enableMocks()

      let lastResponseClone: Response | undefined
      const fetchJestFn = jest.fn<
        Promise<Response>,
        [input: RequestInfo, init?: RequestInit | undefined]
      >((...args) =>
        fetchReal(...args).then((x) => {
          lastResponseClone = x.clone()
          return x
        }),
      )
      fetchMock.mockImplementation(fetchJestFn as any)

      const response = await fetchTokenList()

      expect(new Set(response)).toMatchSnapshot("fetchTokenListResponseSet")
      // fetch gets called once
      expect(fetchJestFn.mock.calls.length).toBe(1)
      // calls the right url
      expect(fetchJestFn.mock.calls[0]).toEqual([
        "https://ropsten-beta-api.zksync.io/api/v0.2/tokens?limit=100&from=latest&direction=older",
      ])
      // gets correct api response
      const fetchResponse: Response = await fetchJestFn.mock.results[0].value
      expect(fetchResponse.status).toBe(200)
      expect((await lastResponseClone?.json()).result).toMatchSnapshot(
        "fetchTokenList",
      )

      fetchMock.disableMocks()
    })
  })
})
