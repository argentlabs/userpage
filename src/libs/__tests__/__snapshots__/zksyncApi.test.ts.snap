// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[int] ZKSYNC Api [int] fetchConfig fetches: fetchConfig 1`] = `
Object {
  "contract": "0x3a49f0f4cf80992976625e1af168f31a12ab5004",
  "depositConfirmations": 0,
  "govContract": "0x80ddc394aa5eeaf20167156a0a1e547e6e42a067",
  "network": "ropsten",
  "zksyncVersion": "contractV4",
}
`;

exports[`[int] ZKSYNC Api [int] fetchTokenList fetches: fetchTokenList 1`] = `
Object {
  "list": Array [
    Object {
      "address": "0x3fd7bf4b341b724d3714dd562b47e49c089e0701",
      "decimals": 18,
      "enabledForFees": true,
      "id": 11,
      "symbol": "ERC20-11",
    },
    Object {
      "address": "0xb36b2e278f5f1980631ad10f693ab3e1bebd9f70",
      "decimals": 18,
      "enabledForFees": true,
      "id": 10,
      "symbol": "MLTT",
    },
    Object {
      "address": "0x1b46bd2fc40030b6959a2d407f7d16f66afadd52",
      "decimals": 18,
      "enabledForFees": true,
      "id": 9,
      "symbol": "BAT",
    },
    Object {
      "address": "0xc865bcbe4b6ef4b58a790052f2b51b4f06f586ac",
      "decimals": 18,
      "enabledForFees": true,
      "id": 8,
      "symbol": "ZRX",
    },
    Object {
      "address": "0x6856ec11f56267e3326f536d0e9f36ec7f7d1498",
      "decimals": 18,
      "enabledForFees": true,
      "id": 7,
      "symbol": "TUSD",
    },
    Object {
      "address": "0x16c550a97ad2ae12c0c8cf1cc3f8db4e0c45238f",
      "decimals": 18,
      "enabledForFees": true,
      "id": 6,
      "symbol": "BUSD",
    },
    Object {
      "address": "0x7e317ceaa15fe7d5474349043332319c5f28cc11",
      "decimals": 18,
      "enabledForFees": true,
      "id": 5,
      "symbol": "FSN",
    },
    Object {
      "address": "0x5ae45f7f17f0df0b24abe25a5522a9c9341ac04d",
      "decimals": 18,
      "enabledForFees": true,
      "id": 4,
      "symbol": "OKB",
    },
    Object {
      "address": "0x793f38ae147852c37071684cdffc1ff7c87f7d07",
      "decimals": 18,
      "enabledForFees": true,
      "id": 3,
      "symbol": "LINK",
    },
    Object {
      "address": "0x351714df444b8213f2c46aaa28829fc5a8921304",
      "decimals": 18,
      "enabledForFees": true,
      "id": 2,
      "symbol": "DAI",
    },
    Object {
      "address": "0x0000000000000000000000000000000000000000",
      "decimals": 18,
      "enabledForFees": true,
      "id": 0,
      "symbol": "ETH",
    },
  ],
  "pagination": Object {
    "count": 11,
    "direction": "older",
    "from": 11,
    "limit": 100,
  },
}
`;

exports[`[int] ZKSYNC Api [int] fetchTokenList fetches: fetchTokenListResponseSet 1`] = `
Set {
  Object {
    "address": "0x0000000000000000000000000000000000000000",
    "decimals": 18,
    "enabledForFees": true,
    "id": 0,
    "symbol": "ETH",
  },
  Object {
    "address": "0x351714df444b8213f2c46aaa28829fc5a8921304",
    "decimals": 18,
    "enabledForFees": true,
    "id": 2,
    "symbol": "DAI",
  },
  Object {
    "address": "0x793f38ae147852c37071684cdffc1ff7c87f7d07",
    "decimals": 18,
    "enabledForFees": true,
    "id": 3,
    "symbol": "LINK",
  },
  Object {
    "address": "0x5ae45f7f17f0df0b24abe25a5522a9c9341ac04d",
    "decimals": 18,
    "enabledForFees": true,
    "id": 4,
    "symbol": "OKB",
  },
  Object {
    "address": "0x7e317ceaa15fe7d5474349043332319c5f28cc11",
    "decimals": 18,
    "enabledForFees": true,
    "id": 5,
    "symbol": "FSN",
  },
  Object {
    "address": "0x16c550a97ad2ae12c0c8cf1cc3f8db4e0c45238f",
    "decimals": 18,
    "enabledForFees": true,
    "id": 6,
    "symbol": "BUSD",
  },
  Object {
    "address": "0x6856ec11f56267e3326f536d0e9f36ec7f7d1498",
    "decimals": 18,
    "enabledForFees": true,
    "id": 7,
    "symbol": "TUSD",
  },
  Object {
    "address": "0xc865bcbe4b6ef4b58a790052f2b51b4f06f586ac",
    "decimals": 18,
    "enabledForFees": true,
    "id": 8,
    "symbol": "ZRX",
  },
  Object {
    "address": "0x1b46bd2fc40030b6959a2d407f7d16f66afadd52",
    "decimals": 18,
    "enabledForFees": true,
    "id": 9,
    "symbol": "BAT",
  },
  Object {
    "address": "0xb36b2e278f5f1980631ad10f693ab3e1bebd9f70",
    "decimals": 18,
    "enabledForFees": true,
    "id": 10,
    "symbol": "MLTT",
  },
  Object {
    "address": "0x3fd7bf4b341b724d3714dd562b47e49c089e0701",
    "decimals": 18,
    "enabledForFees": true,
    "id": 11,
    "symbol": "ERC20-11",
  },
}
`;
